provider-service/base/pkg/streams/sources/workflow_source.go

```diff
@@ -0,0 +1,154 @@
+package sources
+
+import (
+	"context"
+	"fmt"
+	"github.com/argoproj/argo-workflows/v3/pkg/apis/workflow"
+	"github.com/go-logr/logr"
+	"github.com/sky-uk/kfp-operator/argo/common"
+	. "github.com/sky-uk/kfp-operator/provider-service/base/pkg"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/tools/cache"
+	"os"
+	"strconv"
+	"strings"
+)
+
+var (
+	argoWorkflowsGvr = schema.GroupVersionResource{
+		Group:    workflow.Group,
+		Version:  workflow.Version,
+		Resource: workflow.WorkflowPlural,
+	}
+)
+
+const (
+	pipelineRunIdLabel           = "pipeline/runid"
+	workflowUpdateTriggeredLabel = "pipelines.kubeflow.org/events-published"
+)
+
+type WorkflowSource struct {
+	K8sClient K8sClient
+	Logger    logr.Logger
+	out       chan StreamMessage[*unstructured.Unstructured]
+}
+
+func (ws *WorkflowSource) Out() <-chan StreamMessage[*unstructured.Unstructured] {
+	return ws.out
+}
+
+func NewWorkflowSource(ctx context.Context, namespace string, k8sClient K8sClient) (*WorkflowSource, error) {
+	logger := common.LoggerFromContext(ctx)
+
+	workflowSource := &WorkflowSource{
+		K8sClient: k8sClient,
+		Logger:    logger,
+		out:       make(chan StreamMessage[*unstructured.Unstructured]),
+	}
+
+	go func() {
+		if err := workflowSource.start(ctx, namespace); err != nil {
+			logger.Error(err, "failed to start KFP event source")
+			os.Exit(1)
+		}
+	}()
+
+	return workflowSource, nil
+}
+
+func (ws *WorkflowSource) start(ctx context.Context, namespace string) error {
+	ws.Logger.Info("starting workflow event data source...")
+	kfpOperatorExistsRequirement, err := labels.NewRequirement(
+		pipelineRunIdLabel,
+		selection.Exists,
+		[]string{},
+	)
+	if err != nil {
+		ws.Logger.Error(err, "failed to construct requirement")
+		return err
+	}
+	workflowUpdateTriggeredRequirement, err := labels.NewRequirement(
+		workflowUpdateTriggeredLabel,
+		selection.NotEquals,
+		[]string{strconv.FormatBool(true)},
+	)
+	if err != nil {
+		ws.Logger.Error(err, "failed to construct requirement")
+		return err
+	}
+
+	workflowListOptions := metav1.ListOptions{
+		LabelSelector: labels.NewSelector().
+			Add(*kfpOperatorExistsRequirement).
+			Add(*workflowUpdateTriggeredRequirement).
+			String(),
+	}
+
+	factory := dynamicinformer.NewFilteredDynamicSharedInformerFactory(
+		ws.K8sClient.Client,
+		0,
```

Comment:
magic number? might be worth clarifying what this is (named parameters?)

Reply:
golang function calls don't support named parameters

Reply:
created a constant for it.

Reply:
changed in 847762891ac9e4c53e697be2fbd16d03bfb70260